### PR TITLE
[ttl] Fix convert-ttl-to-compute CB push/wait inversion for re-consumed outputs (#666)

### LINKED ISSUE #666: [BUG] Compute deadlock: loop-carried recurrence + compiler-managed intermediate DFBs
## Summary

A `ttl.compute` kernel that combines a **loop-carried recurrence** (e.g. a
running `max` updated each iteration) with **compiler-managed intermediate
values** (matmul / reduce / transpose results left as SSA, materialized by
`ttl-insert-intermediate-dfbs`) **deadlocks on device**. Removing only the
loop-carried recurrence — keeping the same matmul/reduce/exp/transpose SSA
chain — makes the identical kernel pass.

Reproduces on `main` (commit 8eb21608) on Blackhole.

## Reproducers (complete source)

Single file, two ops selectable via the `REPRO` env var. `v2_reuse` (no
loop-carried state) **PASSES**; `v3_state` (= v2 **plus** one running-max
recurrence `s = max(s_old, reduce_max(qk))`) **HANGS**. That single added
recurrence is the entire delta.

```python
# _repro_grow.py
import os
import torch
import ttl
import ttnn

TILE = 32
N_ITERS = 4
K = 2
M = 2


def to_dram(dev, t):
    return ttnn.from_torch(
        t.contiguous(), dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT,
        device=dev, memory_config=ttnn.DRAM_MEMORY_CONFIG,
    )


@ttl.operation(grid=(1, 1))
def v2_reuse(a, b, c, out):
    a_cb = ttl.make_dataflow_buffer_like(a, shape=(1, K), block_count=2)
    b_cb = ttl.make_dataflow_buffer_like(b, shape=(M, K), block_count=2)
    qk_cb = ttl.make_dataflow_buffer_like(a, shape=(1, M), block_count=2)
    mx_cb = ttl.make_dataflow_buffer_like(a, shape=(1, 1), block_count=2)
    es_cb = ttl.make_dataflow_buffer_like(a, shape=(1, 1), block_count=2)
    out_cb = ttl.make_dataflow_buffer_like(out, shape=(1, 1), block_count=N_ITERS)

    @ttl.compute()
    def compute():
        for _ in range(N_ITERS):
            a_blk = a_cb.wait()
            b_blk = b_cb.wait()
            qk_w = qk_cb.reserve(); qk_w.store(a_blk @ ttl.transpose(b_blk))
            qk = qk_cb.wait()
            mx_w = mx_cb.reserve(); mx_w.store(ttl.math.reduce_max(qk, dims=[1]))
            mx = mx_cb.wait()
            mx_bc = ttl.block.broadcast(mx, dims=[1], shape=(1, M))
            ex = ttl.exp(ttl.sub(qk, mx_bc))
            es_w = es_cb.reserve(); es_w.store(ttl.math.reduce_sum(ex, dims=[1]))
            es = es_cb.wait()
            ow = out_cb.reserve(); ow.store(es)

    @ttl.datamovement()
    def dm():
        for _ in range(N_ITERS):
            ad = a_cb.reserve(); ttl.copy(a[0:1, 0:K], ad)
            bd = b_cb.reserve(); ttl.copy(b[0:M, 0:K], bd)
        for _ in range(N_ITERS):
            ob = out_cb.wait(); ttl.copy(ob, out[0:1, 0:1])

    @ttl.datamovement()
    def dm_unused():
        pass


@ttl.operation(grid=(1, 1))
def v3_state(a, b, c, out):
    a_cb = ttl.make_dataflow_buffer_like(a, shape=(1, K), block_count=2)
    b_cb = ttl.make_dataflow_buffer_like(b, shape=(M, K), block_count=2)
    qk_cb = ttl.make_dataflow_buffer_like(a, shape=(1, M), block_count=2)
    mx_cb = ttl.make_dataflow_buffer_like(a, shape=(1, 1), block_count=2)
    es_cb = ttl.make_dataflow_buffer_like(a, shape=(1, 1), block_count=2)
    st_cb = ttl.make_dataflow_buffer_like(a, shape=(1, 1), block_count=2)  # loop-carried state
    out_cb = ttl.make_dataflow_buffer_like(out, shape=(1, 1), block_count=N_ITERS)

    @ttl.compute()
    def compute():
        s0 = st_cb.reserve(); s0.store(ttl.block.fill(-1e30, shape=(1, 1)))
        for _ in range(N_ITERS):
            a_blk = a_cb.wait()
            b_blk = b_cb.wait()
            qk_w = qk_cb.reserve(); qk_w.store(a_blk @ ttl.transpose(b_blk))
            qk = qk_cb.wait()
            mx_w = mx_cb.reserve(); mx_w.store(ttl.math.reduce_max(qk, dims=[1]))
            mx = mx_cb.wait()
            s_old = st_cb.wait()                          # <-- recurrence
            s_new = ttl.math.max(s_old, mx)               # <-- recurrence
            s_next = st_cb.reserve(); s_next.store(s_new)  # <-- recurrence
            mx_bc = ttl.block.broadcast(s_new, dims=[1], shape=(1, M))
            ex = ttl.exp(ttl.sub(qk, mx_bc))
            es_w = es_cb.reserve(); es_w.store(ttl.math.reduce_sum(ex, dims=[1]))
            es = es_cb.wait()
            ow = out_cb.reserve(); ow.store(es)
        _ = st_cb.wait()

    @ttl.datamovement()
    def dm():
        for _ in range(N_ITERS):
            ad = a_cb.reserve(); ttl.copy(a[0:1, 0:K], ad)
            bd = b_cb.reserve(); ttl.copy(b[0:M, 0:K], bd)
        for _ in range(N_ITERS):
            ob = out_cb.wait(); ttl.copy(ob, out[0:1, 0:1])

    @ttl.datamovement()
    def dm_unused():
        pass


def main():
    which = os.getenv("REPRO", "v2_reuse")
    fn = {"v2_reuse": v2_reuse, "v3_state": v3_state}[which]
    dev = ttnn.open_device(device_id=0)
    try:
        a = to_dram(dev, torch.randn(TILE, K * TILE, dtype=torch.bfloat16))
        b = to_dram(dev, torch.randn(M * TILE, K * TILE, dtype=torch.bfloat16))
        c = to_dram(dev, torch.randn(M * TILE, TILE, dtype=torch.bfloat16))
        out = to_dram(dev, torch.zeros(TILE, TILE, dtype=torch.bfloat16))
        print(f"running {which}...")
        fn(a, b, c, out)
        ttnn.synchronize_device(dev)
        print(f"REPRO {which} PASS")
    finally:
        ttnn.close_device(dev)


if __name__ == "__main__":
    main()
```

Run:

```
REPRO=v2_reuse python _repro_grow.py   # prints "REPRO v2_reuse PASS"
REPRO=v3_state python _repro_grow.py   # hangs (no output after "Compiled kernel ready")
```

## Observed behavior

- `v2_reuse`: completes, `REPRO v2_reuse PASS`.
- `v3_state`: kernels compile and dispatch, then the device hangs forever
  (with `TT_METAL_WATCHER=10` the device keeps being polled; no progress).

## Evidence / analysis

- **Not a circular-buffer logic bug.** The IR after `ttl-insert-cb-sync` is
  balanced: every `cb_reserve`/`cb_push` and `cb_wait`/`cb_pop` pairs up
  correctly across the producer/consumer threads.
- **Not back-pressure depth.** Raising the cross-thread DFB `block_count`
  (k/v) from 2 to 4 does **not** fix it.
- **Watcher** (Blackhole) on the single active worker core shows the
  data-movement RISC (NCRISC) blocked at `CRBW` (`cb_reserve_back` wait); the
  compute TRISC waypoints are at *Done* markers (`UABD`, `MWDD`), i.e. the
  compute pipeline is stalled and back-pressures the producer.
- After `ttl-assign-dst`, the loop body is split into many `ttl.dst_section`s
  (one per compute op). The recurrence adds a value (`s_new`) that is
  multi-used — materialized into a compiler-allocated DFB for the broadcast
  **and** written back to the state DFB — which appears to tip the
  DST-section / unpack-math-pack scheduling into a deadlocking order.

Conclusion: the deadlock is in DST-register / `ttl.dst_section` scheduling
(`ttl-assign-dst` -> `ttl-lower-to-loops`), surfaced by combining a
loop-carried recurrence with compiler-managed intermediate DFBs.

## Workaround

Manually stage **every** intermediate in its own single-use DFB instead of
relying on `ttl-insert-intermediate-dfbs` (this keeps DST live-ranges short).
A fully hand-staged flash-attention `flash_shard_kd` (~25 DFBs) runs correctly;
reducing it to rely on compiler-managed intermediates reintroduces this hang.

## Impact

Blocks authoring minimal-DFB online-softmax / flash-attention-style kernels:
any loop-carried recurrence (running max/sum/output) combined with
compiler-materialized op intermediates can deadlock, forcing users to
hand-stage all intermediates.

### PR DESCRIPTION
**Pull request by brnorris03**
Fixes #666.

### Fix
`convert-ttl-to-compute` sinks each multi-output store's full release/consumer run (`cb_push`, the re-consume `cb_wait` and its `attach_cb` users, `cb_pop`) past the fused compute, preserving order; producer `cb_reserve`s are never moved. This restores `cb_push` ≺ `cb_wait` for a re-consumed intra-thread DFB.

Note, this is a temporary fix. TODO(#668) tracks the real follow-up — deferring intermediate-DFB materialization to after fusion (extends #651's `DFBMaterialization`) — which deletes this relocation entirely.

## Tests
- Lit: multi-output compute with a re-consumed output asserts `cb_push` before `cb_wait`.
- Device: `test_recurrence_multi_output_dfb.py` (running-max recurrence, bf16 + fp32).
- `test/TESTING.md`: the bisect-by-reproducer technique used to find this, starting from `examples/test_flash_chain_8node_min_dfb.py` (whose `flash_shard_kd` now runs).

### CODE REVIEW

**Review by vtsilytskyiTT — APPROVED** [review]
LGTM

**Comment by zoecarver** [comment]
I don't think this is the correct approach. So as I understand it, the issue is that insert-cb-sync anchors the push to the store, but fusion pulls these ops into compute, and now the push is in the wrong place. So we are fixing IR that is not correct by construction. 

I wonder, would it be simpler to just move the insert-cb-sync after fusion? I know you have a follow up that defers intermediate DFB materialization after fusion, so I guess that does the same thing. 

But I am hesitant to add this logic to fix IR that's not correct to start, I think it would be better to fix this upstream. 

Is this blocking something such that we want a temporary fix rather than the longer term one?

**Comment by brnorris03** [comment]
Don't think it's an immediate blocker, but could come up again in some situations. Don't mind waiting until various open PRs (that have nontrivial conflict potential) are done before addressing this properly.

**Comment by brnorris03** [comment]
The proper fix is coming soon in a different PR, so closing this. Will reopen if this becomes urgent for some reason.
